hutch_python/ipython_session_timer.py

```diff
@@ -0,0 +1,108 @@
+"""
+This module modifies an ``ipython`` shell to automatically close if it has been
+idle for a certain number of hours. Each hutch can configure their conf.yml
+file to set a timeout duration. The default duration is 48 hours.
+"""
+
+import time
+from threading import Thread
+
+max_idle_time = 172800.0  # number of seconds in 48 hours
+
+
+def configure_timeout(session_timer):
+    global max_idle_time
+    if isinstance(session_timer, int) and session_timer > 0:
+        max_idle_time = session_timer
+
+
+class IPythonSessionTimer:
+    '''
+    Class tracks the amount of time the current `InteractiveShell` instance (henceforth
+    called 'user session') has been idle and closes the session if more than 48
+    hours have passed.
+
+    Time is in seconds (floating point) since the epoch began. (In UNIX the
+    epoch started on January 1, 1970, 00:00:00 UTC)
+
+    Parameters
+    ----------
+    ipython : ``IPython.terminal.interactiveshell.TerminalInteractiveShell``
+        The active ``ipython`` ``Shell``, perhaps the one returned by
+        ``IPython.get_ipython()``.
+
+    Attributes
+    ----------
+    curr_time: float
+        The current time in seconds.
+
+    max_idle_time: float
+        The maximum number of seconds a user session can be idle (default is
+        172800.0 seconds or 48 hours).
+
+    last_active_time: float
+        The time of the last user activity in this session.
+
+    idle_time: float
+        The amount of time the user session has been idle.
+    '''
+
+    def __init__(self, ipython):
+        global max_idle_time
```

Comment:
nitpick 1: no need to declare a global if you don't need to modify it. The scope here can read the variable with no issues without the `global` keyword.
```suggestion
```